hidpp20: implement all of REPROG_CONTROLS_V4

Vtec234 opened the pull request:
Please see the discussion in #733.

Klap-in commented on lib/logitech_receiver/hidpp20.py:
```diff
@@ -357,53 +359,276 @@ def __len__(self):
+
+class ReprogrammableKeyV4(ReprogrammableKey):
+    """Information about a control present on a device with the `REPROG_CONTROLS_V4` feature.
+    Ref: https://lekensteyn.nl/files/logitech/x1b04_specialkeysmsebuttons.html
```
Hi,
This looks like version 2 of the x1b04 specification. Version 4 is available at https://drive.google.com/file/d/10imcbmoxTJ1N510poGdsviEhoFfB_Ua4/view (according to changelog two additional capabilities are described)
(For more docs see https://github.com/Logitech/cpg-docs (work in progress) and https://drive.google.com/drive/u/0/folders/0BxbRzx7vEV7eWmgwazJ3NUFfQ28)

Vtec234 replied:
Thanks! I will have a look, especially `forceRawXY` looks relevant.

pfps replied:
I think that all pointers to Logitech documents should be to the Google drive unless there is a version available at the Logitech Github repository.

Vtec234 commented:
I verified that this produces the exact same list of reprogrammable controls as the current code on my device, but of course that's just one mouse. It would be good if people with other devices (keyboards especially) could check that everything still works.

pfps commented:
Yes, that's one of the things I'm going to check out.

barolo commented:
There's no UI for this yet correct? it only shows for my MX Master 2S, it doesn't for Craft which also has this feature

pfps commented:
So far this is some upgrades to internal data structures.   The only place that might look different is in `solaar show` (I haven't checked to see if there are any changes there yet).
So far Solaar only makes visible the ability to change the action for a key or button to a different one.  This ability is mediated by the group and gmask portions of the feature.   The Craft has group 0 and an empty gmask, which means that the keys cannot be changed in the way Solaar currently supports.  Support for PERSISTENT REMAPPABLE ACTION {1C00} is in the works, and thus will likely allow reprogramming of keys on the Craft.

barolo commented:
@pfps thanks for the info! happy with all the activity around solaar currently

pfps commented on lib/logitech_receiver/hidpp20.py:
```diff
@@ -357,53 +359,276 @@ def __len__(self):
+            divert = True
+
+        if divert is not None and special_keys.KEY_FLAG.divertable not in self.flags:
+            raise FeatureNotSupported(f'Tried to divert non-divertable control {self.key} ' + f'on device {self._device}.')
```
The string split should probably be undone here and in a few other places.   The Solaar style has been set up for long line lengths (about 120 characters, I think).

FFY00 replied:
The line limit is set for 127. I am not sure if yapf allows the line to be split here, but if it does, please do it.

pfps commented on lib/logitech_receiver/hidpp20.py:
```diff
@@ -357,53 +359,276 @@ def __len__(self):
+            raise FeatureNotSupported(f'Tried to divert non-divertable control {self.key} ' + f'on device {self._device}.')
+        if persist is not None and special_keys.KEY_FLAG.persistently_divertable not in self.flags:
+            raise FeatureNotSupported(
+                'Tried to persistently divert non-persistently-divertable control ' + f'{self.key} on device {self._device}.'
```
f-strings push Solaar to Python 3.6.   I think this is OK, but @FFY00 should also approve.
As well, docs/installation.md needs be changed as it currently says 3.2.

FFY00 replied:
I approve.

FFY00 replied:
But if we do it, we should start using it moving forward. This patch also uses the `'' % ()` format.

If anyone has time, you can replace the current `%` and `.format()` usages with f-strings.

pfps replied:
I'll see if there is a good place to put a short paragraph (or document) on coding style in the Solaar documentation.

pfps replied:
@FFY00 I just noticed that setup.py wants 3.5 but the install document says 3.2.   I'll update both to 3.6.

pfps commented on lib/solaar/cli/show.py:
```diff
@@ -184,13 +183,15 @@ def _print_device(dev):
+                print('        %2d: %-26s, default: %-27s => %-26s' % (k.index, k.key, k.default_task, k.mapped_to))
+                print(
+                    '             %s, pos:%d, group:%1d, group mask:%s' %
+                    (', '.join(k.flags), k.pos, k.group, ', '.join(k.group_mask))
```
Nit:  I think that it would look better with no spaces between the groups in the group mask.

pfps commented on lib/logitech_receiver/settings_templates.py:
```diff
@@ -522,29 +521,12 @@ def _feature_pointer_speed():
+        tgts = k.remappable_to
+        if len(tgts) > 1:
+            choices[k.key] = tgts
+
```
The reason that I didn't use KeysArray here was to cut down on the interactions with the device, but you have fixed that.  Good job!

Vtec234 replied:
Thanks! The array is kept lazy and should only fetch info when necessary. But because of this there is a consistency problem. Namely, the state of `device.keys[i].mapped_to` is going to diverge from the real device state if that key is reprogrammed through `feature_map_choices_dynamic` because the setting uses a raw `feature_request` rather than the `device.keys[i].remap` API. This is actually not an observable problem because `mapped_to` is only accessed once (the first fetch is correct, caching kicks in after) in CLI `show` and not elsewhere. For this reason, I would rather handle this in a separate PR than here.

What I would do is rewrite the setting to just use wrapper APIs everywhere rather than raw `feature_request` calls. This would also handle the related problem fact that `write_prefix_bytes=b'\x00',` clears all remapping bits, so these would similarly diverge in `KeysArray` when cached. Thoughts?

pfps commented:
Another nit:  The output from the the Craft keyboard would look better if there was something other than an empty string for the group mask.

FFY00 commented on lib/logitech_receiver/hidpp20.py:
```diff
@@ -357,53 +359,276 @@ def __len__(self):
+                _log.error(f'Trying to read keys on device {device} which has no REPROG_CONTROLS(_VX) support.')
+            self.keyversion = None
         self.keys = [None] * count
+        # The mapping from Control IDs to their native Task IDs.
+        # For example, Control "Left Button" is mapped to Task "Left Click".
+        # When remapping controls, we point the control we want to remap
+        # at a target Control ID rather than a target Task ID. This has the
+        # effect of performing the native task of the target control,
+        # even if the target itself is also remapped. So remapping
+        # is not recursive.
```
Use long strings for comments that span multiple lines.

```
'''

'''
```

pfps replied:
I think that v4 is downward-compatible with previous versions - earlier versions will have 0 for the extra flags so any software will know that these parts of the feature are not available.

Vtec234 replied:
(Resolved long string comments, but keeping open for v4 discussion.)

To be clear, we are talking about `REPROG_CONTROLS_V4` version 1 vs version 4 (yay naming), right? My take on this is:
- For all `get` functions, we can be backwards compatible by reading `unused` bits which should hopefully be 0, and if the device doesn't send enough bytes because a later spec version added another byte to the response format, we just pretend said byte was NUL.
- Notification events aren't a problem, older spec implementations just won't send them.
- But `setCidReporting` seems problematic. Version 4 seems to have added a fifth byte for `analyticsKeyEvt`. What happens if a device implementing earlier versions gets this byte? Also what happens if we set bits marked `unused` in version 1, because they have some function in version 4?

FFY00 replied:
Yes, newer versions are backward-compatible but they can introduce new fields, if we present values from newer versions to the users, they will have a bad experience on older versions.

Acording to your description, I think we can have a hybrid implementation that is aware from v4 for everything other than `setCidReporting`.

pfps replied:
@FFY00 There are two "versions" here - the feature (REPROG_CONTROLS vs REPROG_CONTROLS_V4) and the feature version (versions 0 through 4 of REPROG_CONTROLS_V4).  Everything in Solaar understands the feature differences  - the data structures keep track of the feature (class ReprogrammableKey vs ReprogrammableKeyV4, setting template checks for REPROG_CONTROLS_V4, etc).  The feature version is also taken care of because there are flags in the Cid structure that tell what can be done with the control and devices that don't support higher feature versions will have these flags false.

pfps commented:
These look like the right changes, but could you squash them into the appropriate earlier commit?  That would end up with a much nicer looking set of commits.  This would require some slightly fancy git footwork.

Vtec234 commented:
I'll try to do that. Just a heads up, I'm afk until August, but going to have a look then.

pfps commented:
@Vtec234 Understood.  I'm going to try to get a release out in the meantime as there are a number of new settings that have been added to Solaar.

pfps commented:
@Vtec234 An revamp of how settings work in Solaar would be very useful.  I tried to make the interface easier to use, but I agree that a better way to proceed would be to provide a more flexible underlying implementation so that it is easy to incorporate complex features like reprogrammable keys.

Vtec234 commented:
I think I addressed all comments, could you please have a look again? I added support for v4 of the `REPROG_CONTROLS_V4` spec which is hopefully backwards-compatible at least down to v2 (although I can't really test this as I don't own older devices). In adding this I also simplified the reporting flags handling a bit.

> An revamp of how settings work in Solaar would be very useful. I tried to make the interface easier to use, but I agree that a better way to proceed would be to provide a more flexible underlying implementation so that it is easy to incorporate complex features like reprogrammable keys.

I agree. I think it would be a decent amount of work to implement.

pfps commented:
@barolo Can you test out this PR?  I think you have devices with reprogrammable keys (even if they only divert so Solaar can't reprogram them).

barolo commented:
@pfps 
What I should be looking for?
solar show:
```
Unifying Receiver
  Device path  : /dev/hidraw0
  USB id       : 046d:c52b
  Serial       : BC36E202
    Firmware   : 24.10.B0036
    Bootloader : 01.08
    Other      : AA.AC
  Has 2 paired device(s) out of a maximum of 6.
  Notifications: wireless, software present (0x000900)
  Device activity counters: 1=158, 2=79

  1: Wireless Mouse MX Master 2S
     Codename     : MX Master 2S
     Kind         : mouse
     Wireless PID : 4069
     Protocol     : HID++ 4.5
     Polling rate : 8 ms (125Hz)
     Serial number: 3A64AEB2
        Bootloader: BOT 56.00.B0003
          Firmware: MPM 12.00.B0003
          Firmware: MPM 12.00.B0003
             Other: 
     The power switch is located on the base.
     Supports 32 HID++ 2.0 features:
         0: ROOT                   {0000}   
         1: FEATURE SET            {0001}   
         2: DEVICE FW VERSION      {0003}   
         3: DEVICE NAME            {0005}   
         4: WIRELESS DEVICE STATUS {1D4B}   
         5: RESET                  {0020}   
         6: BATTERY STATUS         {1000}   
         7: CONFIG DEVICE PROPS    {1806}   internal, hidden
         8: CHANGE HOST            {1814}   
            Change Host: 2:medius
         9: REPROG CONTROLS V4     {1B04}   
            Actions: {'80': 80, '81': 81, '82': 82, '83': 83, '86': 86, '195': 195, '196': 196}
        10: ADJUSTABLE DPI         {2201}   
            Sensitivity (DPI): 1200
        11: VERTICAL SCROLLING     {2100}   
            Roller type: 3G
            Ratchet per turn: 24
            Scroll lines: 0
        12: SMART SHIFT            {2110}   
            Smart Shift: 0
        13: HIRES WHEEL            {2121}   
            Multiplier: 8
            Has invert
              Normal wheel motion
            Has ratchet switch
              Free wheel mode
            High resolution mode
            HID notification
            High Resolution Wheel Invert: False
            Wheel Resolution: True
        14: GESTURE 2              {6501}   
        15: DFUCONTROL UNSIGNED    {00C1}   
        16: DFUCONTROL SIGNED      {00C2}   
        17: unknown:1813           {1813}   internal, hidden
        18: unknown:1830           {1830}   internal, hidden
        19: unknown:1890           {1890}   internal, hidden
        20: unknown:1891           {1891}   internal, hidden
        21: unknown:18A1           {18A1}   internal, hidden
        22: unknown:18C0           {18C0}   internal, hidden
        23: unknown:1DF3           {1DF3}   internal, hidden
        24: unknown:1E00           {1E00}   hidden
        25: unknown:1EB0           {1EB0}   internal, hidden
        26: unknown:1803           {1803}   internal, hidden
        27: unknown:1861           {1861}   internal, hidden
        28: unknown:9001           {9001}   internal, hidden
        29: unknown:9200           {9200}   internal, hidden
        30: unknown:9240           {9240}   internal, hidden
        31: OOBSTATE               {1805}   internal, hidden
     Has 8 reprogrammable keys:
         0: Left Button               , default: Left Click                  => Left Click                
             mse, pos:0, group:1, group mask:g1
             reporting: default
         1: Right Button              , default: Right Click                 => Right Click               
             mse, pos:0, group:1, group mask:g1
             reporting: default
         2: Middle Button             , default: Mouse Middle Button         => Mouse Middle Button       
             mse, reprogrammable, divertable, raw XY, pos:0, group:3, group mask:g1,g2,g3
             reporting: default
         3: Back Button               , default: Mouse Back Button           => Mouse Back Button         
             mse, reprogrammable, divertable, raw XY, pos:0, group:2, group mask:g1,g2
             reporting: default
         4: Forward Button            , default: Mouse Forward Button        => Mouse Forward Button      
             mse, reprogrammable, divertable, raw XY, pos:0, group:2, group mask:g1,g2
             reporting: default
         5: App Switch Gesture        , default: Gesture Button Navigation   => Gesture Button Navigation 
             mse, reprogrammable, divertable, raw XY, pos:0, group:3, group mask:g1,g2,g3
             reporting: default
         6: Smart Shift               , default: Smart Shift                 => Smart Shift               
             mse, reprogrammable, divertable, raw XY, pos:0, group:3, group mask:g1,g2,g3
             reporting: default
         7: Virtual Gesture Button    , default: Virtual Gesture Button      => Virtual Gesture Button    
             divertable, virtual, raw XY, force raw XY, pos:0, group:4, group mask:empty
             reporting: default
     Battery: 20%, discharging, next level 5%.

  2: Craft Advanced Keyboard
     Codename     : Craft
     Kind         : keyboard
     Wireless PID : 4066
     Protocol     : HID++ 4.5
     Polling rate : 8 ms (125Hz)
     Serial number: 40163908
        Bootloader: BOT 41.00.B0014
          Firmware: MPK 07.01.B0015
             Other: 
             Other: 
     The power switch is located on the edge of top right corner.
     Supports 39 HID++ 2.0 features:
         0: ROOT                   {0000}   
         1: FEATURE SET            {0001}   
         2: DEVICE FW VERSION      {0003}   
         3: DEVICE NAME            {0005}   
         4: WIRELESS DEVICE STATUS {1D4B}   
         5: RESET                  {0020}   
         6: DEVICE FRIENDLY NAME   {0007}   
         7: BATTERY STATUS         {1000}   
         8: CHANGE HOST            {1814}   
            Change Host: 2:DESKTOP-SL4D3O2
         9: HOSTS INFO             {1815}   
            Host 0 (paired): DESKTOP-SL4D3O2
            Host 1 (paired): DESKTOP-SL4D3O2
            Host 2 (paired): MacBook
        10: BACKLIGHT2             {1982}   
            Backlight: True
        11: REPROG CONTROLS V4     {1B04}   
        12: PERSISTENT REMAPPABLE ACTION {1C00}   
        13: K375S FN INVERSION     {40A3}   
            Swap Fx function: False
        14: ENCRYPTION             {4100}   
        15: LOCK KEY STATE         {4220}   
        16: KEYBOARD DISABLE KEYS  {4521}   
            Disable keys: {'1': True, '2': False, '4': True, '8': True, '16': False}
        17: MULTIPLATFORM          {4531}   
            Set OS: Windows
        18: CROWN                  {4600}   
        19: DFUCONTROL SIGNED      {00C2}   
        20: unknown:1803           {1803}   internal, hidden
        21: CONFIG DEVICE PROPS    {1806}   internal, hidden
        22: unknown:1813           {1813}   internal, hidden
        23: OOBSTATE               {1805}   internal, hidden
        24: unknown:1830           {1830}   internal, hidden
        25: unknown:1890           {1890}   internal, hidden
        26: unknown:1891           {1891}   internal, hidden
        27: unknown:1801           {1801}   internal, hidden
        28: unknown:18A1           {18A1}   internal, hidden
        29: unknown:9280           {9280}   internal, hidden
        30: unknown:1A20           {1A20}   internal, hidden
        31: unknown:1DF3           {1DF3}   internal, hidden
        32: unknown:1E00           {1E00}   hidden
        33: unknown:1EB0           {1EB0}   internal, hidden
        34: unknown:1861           {1861}   internal, hidden
        35: unknown:18B0           {18B0}   internal, hidden
        36: unknown:92C0           {92C0}   internal, hidden
        37: unknown:9203           {9203}   internal, hidden
        38: unknown:3615           {3615}   internal, hidden
     Has 24 reprogrammable keys:
         0: Host Switch channel 1     , default: HostSwitch Channel 1        => HostSwitch Channel 1      
             nonstandard, divertable, persistently divertable, pos:0, group:0, group mask:empty
             reporting: default
         1: Host Switch channel 2     , default: HostSwitch Channel 2        => HostSwitch Channel 2      
             nonstandard, divertable, persistently divertable, pos:0, group:0, group mask:empty
             reporting: default
         2: Host Switch channel 3     , default: HostSwitch Channel 3        => HostSwitch Channel 3      
             nonstandard, divertable, persistently divertable, pos:0, group:0, group mask:empty
             reporting: default
         3: Brightness Down           , default: Brightness Down             => Brightness Down           
             is FN, FN sensitive, reprogrammable, divertable, persistently divertable, pos:1, group:0, group mask:empty
             reporting: default
         4: Brightness Up             , default: Brightness Up               => Brightness Up             
             is FN, FN sensitive, reprogrammable, divertable, persistently divertable, pos:2, group:0, group mask:empty
             reporting: default
         5: Mission Control/Task View , default: Mission Control/Task View   => Mission Control/Task View 
             is FN, FN sensitive, reprogrammable, divertable, persistently divertable, pos:3, group:0, group mask:empty
             reporting: default
         6: Dashboard Launchpad/Action Center, default: Dashboard Launchpad/Action Center => Dashboard Launchpad/Action Center
             is FN, FN sensitive, reprogrammable, divertable, persistently divertable, pos:4, group:0, group mask:empty
             reporting: default
         7: Show Desktop              , default: Show Desktop                => Show Desktop              
             is FN, FN sensitive, reprogrammable, divertable, persistently divertable, pos:5, group:0, group mask:empty
             reporting: default
         8: Backlight Down            , default: Backlight Down              => Backlight Down            
             is FN, FN sensitive, reprogrammable, divertable, persistently divertable, pos:6, group:0, group mask:empty
             reporting: default
         9: Backlight Up              , default: Backlight Up                => Backlight Up              
             is FN, FN sensitive, reprogrammable, divertable, persistently divertable, pos:7, group:0, group mask:empty
             reporting: default
        10: Previous Fn               , default: Previous                    => Previous                  
             is FN, FN sensitive, reprogrammable, divertable, persistently divertable, pos:8, group:0, group mask:empty
             reporting: default
        11: Play/Pause Fn             , default: Play/Pause                  => Play/Pause                
             is FN, FN sensitive, reprogrammable, divertable, persistently divertable, pos:9, group:0, group mask:empty
             reporting: default
        12: Next Fn                   , default: Next                        => Next                      
             is FN, FN sensitive, reprogrammable, divertable, persistently divertable, pos:10, group:0, group mask:empty
             reporting: default
        13: Mute Fn                   , default: Mute                        => Mute                      
             is FN, FN sensitive, reprogrammable, divertable, persistently divertable, pos:11, group:0, group mask:empty
             reporting: default
        14: Volume Down Fn            , default: Volume Down                 => Volume Down               
             is FN, FN sensitive, reprogrammable, divertable, persistently divertable, pos:12, group:0, group mask:empty
             reporting: default
        15: Volume Up Fn              , default: Volume Up                   => Volume Up                 
             nonstandard, reprogrammable, divertable, persistently divertable, pos:0, group:0, group mask:empty
             reporting: default
        16: Calculator                , default: Calculator                  => Calculator                
             nonstandard, reprogrammable, divertable, persistently divertable, pos:0, group:0, group mask:empty
             reporting: default
        17: Screen Capture/Print Screen, default: Screen Capture              => Screen Capture            
             nonstandard, reprogrammable, divertable, persistently divertable, pos:0, group:0, group mask:empty
             reporting: default
        18: App Contextual Menu/Right Click, default: Right Click/App Contextual Menu => Right Click/App Contextual Menu
             nonstandard, reprogrammable, divertable, persistently divertable, pos:0, group:0, group mask:empty
             reporting: default
        19: Lock PC                   , default: WindowsLock                 => WindowsLock               
             nonstandard, reprogrammable, divertable, persistently divertable, pos:0, group:0, group mask:empty
             reporting: default
        20: Left Arrow                , default: Keyboard Left Arrow         => Keyboard Left Arrow       
             nonstandard, divertable, persistently divertable, pos:0, group:0, group mask:empty
             reporting: default
        21: Right Arrow               , default: Keyboard Right Arrow        => Keyboard Right Arrow      
             nonstandard, divertable, persistently divertable, pos:0, group:0, group mask:empty
             reporting: default
        22: F Lock                    , default: Do Nothing One              => Do Nothing One            
             is FN, pos:0, group:0, group mask:empty
             reporting: default
        23: unknown:0034              , default: Do Nothing One              => Do Nothing One            
             nonstandard, pos:0, group:0, group mask:empty
             reporting: default
     Battery: 50%, discharging, next level 20%.
```

pfps commented:
Just that things generally work - `solaar show` which you did; keys with no groups can't be programmed, which you did.  The only thing left is whether you can reprogram one of the buttons on your mouse.

barolo commented:
@pfps They can't be reprogrammed via solaar
